sdk/include/opentelemetry/sdk/trace/sampler.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include "opentelemetry/common/attribute_value.h"
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/sdk/common/atomic_shared_ptr.h"
+#include "opentelemetry/trace/span.h"
+#include "opentelemetry/trace/trace_id.h"
+#include "opentelemetry/trace/tracer.h"
+#include "opentelemetry/version.h"
+
+#include <map>
+#include <memory>
+#include <string>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+namespace trace_api = opentelemetry::trace;
+class Sampler
+{
+private:
+  // Placeholder
+  class SpanContext
+  {};
+
+public:
+  virtual ~Sampler() = default;
+  /**
+   * Called during Span creation to make a sampling decision.
+   *
+   * @param parent_context TODO: a shared pointer of the SpanContext of a parent Span. 
+   *     null if this is a root span.
+   * @param trace_id the TraceId for the new Span. This will be identical to that in
+   *     the parentContext, unless this is a root span.
+   * @param name the name of the new Span.
+   * @param spanKind the trace_api::SpanKind of the Span.
+   * @param attributes TODO: current map is a placeholder.
+   *     list of AttributeValue with their keys.
+   * @param links TODO: Collection of links that will be associated with the Span to be created.
+   * @return sampling result whether span should be sampled or not.
+   * @since 0.1.0
+   */
+
+  virtual SamplingResult ShouldSample(
+      std::shared_ptr<SpanContext> parent_context,
```

Comment:
Why shared pointer? Should this be a constref instead?

Reply:
According to the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#sampler), we must be able to pass `NULL` values.

I'm not sure about that yet, but we might hold `SpanContext` objects in `shared_ptr`s anyway, because a span's context should be available to the span itself and to its children?

Reply:
Thanks for the spec link. The NULL is for cases where there is no parent, right? We could pass a regular pointer if the Sampler isn't allowed to hold on to it.

Reply:
Good call. Let's make this a `const SpanContext* parent_context`.

http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#r30-take-smart-pointers-as-parameters-only-to-explicitly-express-lifetime-semantics

Reply:
`std::shared_ptr<SpanContext>` would make sense if the Sampler would take (shared) ownership of the span context beyond the scope of this call.

This is not the case (the sampler won't store the span context), so a pointer will do.